gym/lib/gym/code_signing_mapping.rb

```diff
@@ -0,0 +1,136 @@
+require 'xcodeproj'
+
+module Gym
+  class CodeSigningMapping
+    attr_accessor :project
+
+    attr_accessor :project_paths
+
+    def initialize(project: nil)
+      self.project = project
+    end
+
+    def merge_profile_mapping(existing_mapping: nil, export_method: nil)
+      final_mapping = existing_mapping.dup # for verbose output at the end of the method
+      project_mapping = self.detect_project_profile_mapping
+
+      # Now it's time to merge the (potentially) existing `provisioningProfiles` of the `export_options`
+      # with the hash we just created. Both might include information about what profile to use
+      # This is important, as :bell: (some users) didn't know to call match for each app target he wants profiles for
+      # before adding this code, we'd only either use whatever we get from match, or what's defined in the Xcode project
+      # With the code below, we'll make sure to take the best of it:
+      #
+      #   1) A provisioning profile is defined in the Xcode project only: we'll take that
+      #   2) A provisioning profile is defined manually from the user: we'll take that (same as 3)
+      #   3) A provisioning profile is defined after calling match: we'll take that (same as 2)
+      #   4) On a conflict (app identifier assigned both in xcode and match)
+      #     4.1) we'll choose whatever matches what's defined in the `export_method`
+      #     4.2) If both include the export_method, we'll prefer the one from match
+      #     4.3) If none has the right export_method, we'll use whatever is defined in the Xcode project
+
+      # 2) is already covered by assigning `final_mapping` above
+      # 3) see 2
```

Comment:
2 and 3 are the same thing, as it's treated the same way, see comment https://github.com/fastlane/fastlane/pull/9952#discussion_r132236924. I'll update the comments to make it more clear and explain more